solidity/contracts/libs/MessageV2.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+import {TypeCasts} from "./TypeCasts.sol";
+
+/**
+ * @title Hyperlane Message Library
+ * @notice Library for formatted messages used by Mailbox
+ **/
+library MessageV2 {
+    using TypeCasts for bytes32;
+
+    /**
+     * @notice Returns formatted (packed) Hyperlane message with provided fields
+     * @dev This function should only be used in memory message construction.
+     * @param _version The version of the origin and destination Mailboxes
+     * @param _nonce A nonce to uniquely identify the message on its origin chain
+     * @param _originDomain Domain of origin chain
+     * @param _sender Address of sender as bytes32
+     * @param _destinationDomain Domain of destination chain
+     * @param _recipient Address of recipient on destination chain as bytes32
+     * @param _messageBody Raw bytes of message body
+     * @return Formatted message
+     */
+    function formatMessage(
+        uint32 _version,
+        uint256 _nonce,
+        uint32 _originDomain,
+        bytes32 _sender,
+        uint32 _destinationDomain,
+        bytes32 _recipient,
+        bytes calldata _messageBody
+    ) internal pure returns (bytes memory) {
+        return
+            abi.encodePacked(
+                _version,
+                _nonce,
+                _originDomain,
+                _sender,
+                _destinationDomain,
+                _recipient,
+                _messageBody
+            );
+    }
+
+    /**
+     * @notice Returns the message ID.
+     * @param _message ABI encoded Hyperlane message.
+     * @return ID of `_message`
+     */
+    function id(bytes memory _message) internal pure returns (bytes32) {
```

Comment:
I think it's necessary for message IDs to be a function of the message content. Otherwise, anyone could set their ISM to just always accept messages and DOS the system by filling up the replay protection dictionary

Reply:
hmm interesting, great point